packages/@lwc/engine-core/src/libs/signal-tracker/index.ts

```diff
@@ -0,0 +1,86 @@
+/*
+ * Copyright (c) 2024, salesforce.com, inc.
+ * All rights reserved.
+ * SPDX-License-Identifier: MIT
+ * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
+ */
+import { isFalse, isUndefined } from '@lwc/shared';
+import { Signal } from '@lwc/signals';
+import { logWarnOnce } from '../../shared/logger';
+
+/**
+ * This map keeps track of objects to signals. There is an assumption that the signal is strongly referenced
+ * on the object which allows the SignalTracker to be garbage collected along with the object.
+ */
+const TargetToSignalTrackerMap: WeakMap<Object, SignalTracker> = new WeakMap();
+
+function getSignalTracker(target: Object) {
+    let signalTracker = TargetToSignalTrackerMap.get(target);
+    if (isUndefined(signalTracker)) {
+        signalTracker = new SignalTracker();
+        TargetToSignalTrackerMap.set(target, signalTracker);
+    }
+    return signalTracker;
+}
+
+export function subscribeToSignal(
+    target: Object,
+    signal: Signal<unknown>,
+    update: CallbackFunction
+) {
+    const signalTracker = getSignalTracker(target);
+    if (isFalse(signalTracker.seen(signal))) {
+        signalTracker.subscribeToSignal(signal, update);
+    }
+}
+
+export function unsubscribeFromSignals(target: Object) {
+    if (TargetToSignalTrackerMap.has(target)) {
+        const signalTracker = getSignalTracker(target);
+        signalTracker.unsubscribeFromSignals();
+        signalTracker.reset();
+    }
+}
+
+type CallbackFunction = () => void;
+
+/**
+ * This class is used to keep track of the signals associated to a given object.
+ * It is used to prevent the LWC engine from subscribing duplicate callbacks multiple times
+ * to the same signal. Additionally, it keeps track of all signal unsubscribe callbacks, handles invoking
+ * them when necessary and discarding them.
+ */
+class SignalTracker {
+    private signalToUnsubscribeMap: Map<Signal<unknown>, CallbackFunction> = new Map();
+
+    seen(signal: Signal<unknown>) {
+        return this.signalToUnsubscribeMap.has(signal);
+    }
+
+    subscribeToSignal(signal: Signal<unknown>, update: CallbackFunction) {
+        try {
+            const unsubscribe = signal.subscribe(update);
+            if (typeof unsubscribe === 'function') {
+                this.signalToUnsubscribeMap.set(signal, unsubscribe);
+            }
+        } catch (err) {
+            logWarnOnce(
+                `Attempted to subscribe to an object that has the shape of a signal but received the following error: ${err}`
+            );
```

Comment:
We should probably have a call to action here, i.e. tell the user what they need to do to resolve the error.

AIUI, this will occur if e.g. a plain object has a shape like `{ value: 'foo' }`, right? Meaning we mistakenly sniff it as a signal?

If that's the case, then I don't know if we even want to warn here. `value` is a pretty common name, and I imagine users may be annoyed by even one warning message, if there's nothing they can do about it. What do other signal-using frameworks do in this case?

Reply:
We're checking `typeof obj.subscribe === 'function'` elsewhere, so that is not the case that'd be hit here. Instead, this would occur if an error threw while inside the `subscribe()` call.

We definitely don't want errors in `subscribe()` to be swallowed entirely. However, I think coercing into a string here might not be the right thing. For example, you'd lose the call stack. On the other hand, a thrown object isn't guaranteed to be an actual error object. So maybe something like this:

```
logWarnOnce(`Attempted to subscribe to...: ${err?.stack ?? err}`);
```

Reply:
@nolanlawson the signal shape should be `{ value: 'foo' , subscribe: () => {} }`.

I'll have to do some more research on how other frameworks handle this situation but I know most of them have their own signal implementations, so they may not run into this issue. 

I'll log the stack trace for now as @divmain  suggested and look into how other frameworks handle this in more detail.